File: bench.py
import numpy as np
import time

import nn_np
import nn_ctypes


def bench(name, module, test_points, train_points):
    st = time.time()
    out = module.nn(test_points, train_points)
    print('%s: %f' % (name, time.time() - st))
    return out


def make_points(num_points, num_dims):
    return np.random.random((num_points, num_dims))


def main(num_train_points=10000, num_test_points=20, num_dims=1024):
    test_points = make_points(num_test_points, num_dims)
    train_points = make_points(num_train_points, num_dims)
    methods = [('Numpy', nn_np), ('Ctypes', nn_ctypes)]
    for x, y in methods:
        print(bench(x, y, test_points, train_points))


if __name__ == '__main__':
    main()

File: bench.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include "nn.h"

double rand_double() {
  return rand() / (double) RAND_MAX;
}

void random_array(double *arr, int len) {
  int i;
  for (i = 0; i < len; i++)
    arr[i] = rand_double();
}

int main() {
  int num_train_points=10000, num_test_points=20, num_dims=1024;
  double *train_points, *test_points;
  int *indeces;
  int i;
  srand(time(NULL));
  indeces = malloc(sizeof *indeces * num_test_points);
  train_points = malloc(sizeof *train_points * num_train_points * num_dims);
  test_points = malloc(sizeof *test_points * num_test_points * num_dims);

  for (i = 0; i < num_train_points; i++)
    random_array(&train_points[i * num_dims], num_dims);

  for (i = 0; i < num_test_points; i++)
    random_array(&test_points[i * num_dims], num_dims);

  nn(test_points, num_test_points, train_points, num_train_points, num_dims, indeces);
  for (i = 0; i < num_test_points; i++)
    printf("%d ", indeces[i]);
  printf("\n");

  free(test_points);
  free(train_points);
  free(indeces);

  return 0;
}
